sqlstats/sqlstats_test.go

```diff
@@ -119,3 +122,90 @@ func TestCollect(t *testing.T) {
 		}
 	}
 }
+
+func TestTracerResetRace(t *testing.T) {
```

Comment:
I believe this test can be simplified a bit to make it easier to read. Specifically:

- We don't need to cancel conditionally; cancelling a completed context is already idempotent.
- We don't need to count error values, only the transaction goroutine reports any errors that need to be reported.
- We can simplify the context-done checks a bit.

```go
func TestTracerResetRace(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()

	tracer := &Tracer{}
	db := sql.OpenDB(sqlite.Connector("file:"+t.TempDir()+"/test.db", nil, tracer))
	defer db.Close()

	if _, err := db.ExecContext(ctx, "CREATE TABLE t (c);"); err != nil {
		t.Fatalf("Create table: %v", err)
	}

	errc := make(chan error)
	go func() {
		defer close(errc)
		defer func() {
			if r := recover(); r != nil {
				ch <- fmt.Errorf("panic: %v", r)
			}
		}()
		for n := 0; ctx.Err() == nil; n++ {
			tx, err := db.BeginTx(ctx, nil)
			if err != nil {
				ch <- err
				return
			} else if _, err := tx.Exec("INSERT INTO t VALUES (?)", n); err != nil {
				tx.Rollback()
				ch <- err
				return
			} else if n%2 == 0 {
				continue
			}
			if err := tx.Commit(); err != nil {
				ch <- err
				return
			}
		}

	}()
	go func() {
		for ctx.Err() == nil {
			tracer.Reset()
		}
	}()

	for err := range errc {
		t.Errorf("Unexpected transaction error: %v", err)
		cancel()
	}
}

```

Reply:
Thanks for those pointers, that's much simpler. My original code certainly does look like a bit of an overreaction to not having errgroup to reach for 😅

Reply:
I almost suggested importing errgroup, but then I realized we probably don't really need it because one of the two goroutines never reports anything 🙂